check_nightly_success/check-nightly-success/check.py

```diff
@@ -0,0 +1,96 @@
+"""Check whether a GHA workflow has run successfully in the last N days."""
+# ruff: noqa: INP001
+
+import argparse
+import itertools
+import os
+import sys
+from datetime import datetime
+
+import requests
+
+# Constants
+GITHUB_TOKEN = os.environ["RAPIDS_GH_TOKEN"]
+GOOD_STATUSES = {"success"}
+
+
+def main(
+    repo: str,
+    repo_owner: str,
+    workflow_id: str,
+    max_days_without_success: int,
+    num_attempts: int = 5,
+) -> bool:
+    """Check whether a GHA workflow has run successfully in the last N days."""
+    headers = {"Authorization": f"token {GITHUB_TOKEN}"}
+    url = f"https://api.github.com/repos/{repo_owner}/{repo}/actions/workflows/{workflow_id}/runs"
+    exceptions = []
+    for _ in range(num_attempts):
+        try:
+            response = requests.get(url, headers=headers, timeout=10)
+            response.raise_for_status()
+            break
+        except requests.RequestException as e:
+            exceptions.append(e)
+    else:
+        sep = "\n\t"
+        msg = (
+            f"Failed to fetch {url} after {num_attempts} attempts with the following "
+            f"errors: {sep}{'{sep}'.join(exceptions)}"
+        )
+        raise RuntimeError(msg)
+
+    runs = response.json()["workflow_runs"]
+    tz = datetime.fromisoformat(runs[0]["run_started_at"]).tzinfo
+    now = datetime.now(tz=tz)
+
+    branch_ok = {}
+    for branch, branch_runs in itertools.groupby(runs, key=lambda r: r["head_branch"]):
+        if branch.startswith("v"):  # Skip tags
+            continue
+
+        branch_ok[branch] = False
+        for run in sorted(branch_runs, key=lambda r: r["run_started_at"], reverse=True):
+            if (
+                now - datetime.fromisoformat(run["run_started_at"])
+            ).days > max_days_without_success:
+                break
+            if run["conclusion"] in GOOD_STATUSES:
+                branch_ok[branch] = True
+                break
+
+    all_success = all(branch_ok.values())
+    if not all_success:
```

Comment:
Maybe just find the failed branches and then use that in the print statement below.
```suggestion
    failed_branches = [k for k, v in branch_ok.items() if not v]
    if failed_branches:
```